include/orb/Orbit.h

```diff
@@ -0,0 +1,391 @@
+     * earth orbit, and has a reasonable time stamp (within 20 years of pan epoch).
+     * The validity check should not reject 
+     * gps readings due to reasonable noise of:
+     * TODO add max expected gps error see issue #372
+     *
+     * Low earth orbit is a Orbit that stays between MINORBITRADIUS and MAXORBITRADIUS.
+     */
+    bool _valid{false};
+
+    /** Returns time since gps epoch (ns).
+     * The Orbit must be not propagating.
+     * 
+     * grav calls: 0 */
+    uint64_t nsgpstime() const{
+        return _ns_gps_time;
+    }
+
+    /** Return position of the sat (m).
+     * The Orbit must be not propagating.
+     * 
+     * grav calls: 0 */
+    lin::Vector3d recef() const{
+        return _recef;
+    }
+
+    /** Return velocity of the sat (m/s).
+     * The Orbit must be not propagating.
+     * 
+     * grav calls: 0 */
+    lin::Vector3d vecef() const{
+        return _vecef;
+    }
+
+    /** Return true if the Orbit is valid.
+     * A valid orbit has finite and real position and velocity, is in low
+     * earth orbit, and has a reasonable time stamp (within 20 years of pan epoch).
+     * The validity check should not reject 
+     * gps readings due to reasonable noise of:
+     * TODO add max expected gps error see issue #372
+     *
+     * Low earth orbit is a Orbit that stays between MINORBITRADIUS and MAXORBITRADIUS.
+     * 
+     * grav calls: 0 */
+    bool valid() const{
+        return _valid;
+    }
+    
+    /** 
+     * Helper function calculate if the Orbit is valid, see valid().
+     * If the orbit is invalid set the orbit to the default values.
+     * 
+     * grav calls: 0 */
+    void _check_validity(){
+        double r2= lin::fro(_recef);
+        //TODO add checks for time and velocity see issue #372
+        //note if position is NAN, these checks will be false.
+        if (r2<MAXORBITRADIUS*MAXORBITRADIUS && r2>MINORBITRADIUS*MINORBITRADIUS){
+            _valid= true;
+        } else {
+            _valid= false;
+            _recef= lin::nans<lin::Vector3d>();
+            _vecef= lin::nans<lin::Vector3d>();
+            _ns_gps_time= 0;
+        }
+    }
+
+    /** Gravity function in International Terrestrial Reference System coordinates.
+     * 
+     * grav calls: 1
+     * @param[in] r_ecef (Above the surface of earth): The location where the gravity is calculated, units m.
+     * @param[out] g_ecef: Acceleration due to gravity, units m/s^2. 
+     * @param[out] potential: Gravity potential, the acceleration is the gradient of this (J/kg).
+     */
+    static void calc_geograv(const lin::Vector3d& r_ecef, lin::Vector3d& g_ecef, double& potential) {
+        geograv::Vector in;
+        geograv::Vector g;
+        in.x= r_ecef(0);
+        in.y= r_ecef(1);
+        in.z= r_ecef(2);
+        potential=geograv::GeoGrav(in, g,PANGRAVITYMODEL,true);
+        g_ecef(0)= g.x;
+        g_ecef(1)= g.y;
+        g_ecef(2)= g.z;
+    }
```

Comment:
Do we want to put this function in the environment models potentially? It's fairly likely we may it outside of just the orbit propagator. Or externally would we use a far less accurate gravity model?

Reply:
Yeah, I am trying to only use this function in the orbit propagator because it is so expensive, but it is static so any other function can call it if needed.